summoner-cli/src/Summoner/Template/GitHub.hs

```diff
@@ -121,49 +121,121 @@ gitHubFiles Settings{..} = concat
+        , "    - uses: actions/setup-haskell" <> ghcActionsSetupHaskellVersion
+        , "      id: setup-haskell-cabal"
         , "      name: Setup Haskell"
         , "      with:"
         , "        ghc-version: ${{ matrix.ghc }}"
         , "        cabal-version: ${{ matrix.cabal }}"
         , ""
-        , "    - uses: actions/cache@v1"
+        , "    - name: Freeze"
+        , "      run: |"
+        , "        cabal v2-freeze"
+        , ""
+        , "    - uses: actions/cache" <> ghcActionsCacheVersion
         , "      name: Cache ~/.cabal/store"
         , "      with:"
-        , "        path: ~/.cabal/store"
-        , "        key: ${{ runner.os }}-${{ matrix.ghc }}-cabal"
+        , "        path: ${{ steps.setup-haskell-cabal.outputs.cabal-store }}"
+        , "        key: ${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('cabal.project.freeze') }}"
         , ""
         , "    - name: Build"
         , "      run: |"
-        , "        " <> cabalUpdate
+        , "        " <> cabalConfigure
         , "        " <> cabalBuild
         , ""
         , "    - name: Test"
         , "      run: |"
         , "        " <> cabalTest
+        , ""
         ]
 
+    ghActionsStack :: [Text]
+    ghActionsStack =
+        [ "  stack:"
+        , "    name: stack / ghc ${{ matrix.ghc }}"
+        , "    runs-on: ubuntu-latest"
+        , "    strategy:"
+        , "      matrix:"
+        , "        stack: [" <> quote defaultStack <> "]"
+        , "        ghc:"
+        ]
+        <> map (indent 10 <>) ghActionsVersions
+        <>
+        [ ""
+        , "    steps:"
+        , "    - uses: actions/checkout" <> ghcActionsCheckoutVersion
+        , "      if: github.event.action == 'opened' || github.event.action == 'synchronize' || github.event.ref == 'refs/heads/master'"
+        , ""
+        , "    - uses: actions/setup-haskell" <> ghcActionsSetupHaskellVersion
+        , "      name: Setup Haskell Stack"
+        , "      with:"
+        , "        ghc-version: ${{ matrix.ghc }}"
+        , "        stack-version: ${{ matrix.stack }}"
+        , ""
+        , "    - uses: actions/cache" <> ghcActionsCacheVersion
+        , "      name: Cache ~/.stack"
+        , "      with:"
+        , "        path: ~/.stack"
+        , "        key: ${{ runner.os }}-${{ matrix.ghc }}-stack"
+        , ""
+        , "    - name: Build"
+        , "      run: |"
+        , "        " <> stackBuild
+        , ""
+        , "    - name: Test"
+        , "      run: |"
+        , "        " <> stackTest
+        ]
 
     ghActionsVersions :: [Text]
     ghActionsVersions = map
         (\ghc -> "- " <> quote (showGhcVer ghc))
         settingsTestedVersions
 
+    ghActionsExcludes :: [Text]
+    ghActionsExcludes =
+        let latestVersion = viaNonEmpty last $ sort settingsTestedVersions
+            versionsToExclude = filter (\version -> pure version /= latestVersion) settingsTestedVersions
```

Comment:
If we would pattern just pattern match on the sorted list from the previous line, you could get the head, and the rest at once, without need to filter and walk the list again 🙂

Reply:
If we decide to go with sort in descending order I will do that :) If we decide to go with `defaultGhc` I will get rid of sorting and leave just the `filter` call.

Reply:
`filter` removed